### PR TITLE
Entra ID integration: add onboarding script

### PR DESCRIPTION
**Pull request by justinas**
Adds a new `teleport integration configure` subcommand to set up Azure OIDC integration. The command sets up:
* An Entra ID enterprise application, both as an SSO connector to allow login with Entra to Teleport, and as a way for us to access Entra ID directory programatically (via OIDC auth).
* Optionally, when access graph support is requested, queries information about existing enterprise applications from the "private" Azure API ([see RFD](https://github.com/gravitational/access-graph/blob/rfd/0006/rfd/0006-entra-id-sync.md#private-azure-api) for more info) and produces a JSON payload that the user is expected to upload when finishing the integration onboarding in the UI.

Also adds the web endpoint in the Proxy that generates the script wrapper for this.

Still missing:
* Frontend to onboard Entra that utilizes this. Will come as a subsequent PR.

### CODE REVIEW

File: lib/integrations/azureoidc/accessgraph.go [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+func CreateTAGCacheFile(ctx context.Context) error {
+	graphClient, err := createGraphClient()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Get information about enterprise apps
+	appResp, err := graphClient.Applications().Get(ctx, nil)
+	if err != nil {
+		panic(err)
```
Comment (jakule) [inline lib/integrations/azureoidc/accessgraph.go:79]: I don't think panicking here is a good idea.

File: lib/integrations/azureoidc/accessgraph.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+// gzipBytes compresses the given byte slice, returning the result as a new byte slice.
+func gzipBytes(src []byte) []byte {
+	out := new(bytes.Buffer)
+	writer := gzip.NewWriter(out)
+
+	_, err := io.Copy(writer, bytes.NewReader(src))
+	// We do not expect in-memory bytes I/O to fail.
+	if err != nil {
+		panic(err)
```
Comment (jakule) [inline lib/integrations/azureoidc/accessgraph.go:146]: I'd still return the error here. The chances are close to 0, but we should not use panics in non critical code (unrecoverable failure that should stop the whole app).

File: lib/integrations/azureoidc/enterprise_app.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		return appID, tenantID, trace.Wrap(err)
+	}
+
+	graphClient, err := createGraphClient()
+	if err != nil {
+		return appID, tenantID, trace.Wrap(err)
+	}
+
+	displayName := "Teleport" + " " + strings.TrimPrefix(proxyPublicAddr, "https://")
```
Comment (jakule) [inline lib/integrations/azureoidc/enterprise_app.go:68]: nit: You can use `url.Url` to extract the hostname and make sure that the output is correct.

File: lib/integrations/azureoidc/enterprise_app.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	configuration := &serviceprincipals.ServicePrincipalsRequestBuilderGetRequestConfiguration{
+		QueryParameters: requestParameters,
+	}
+	spResponse, err := graphClient.ServicePrincipals().Get(ctx, configuration)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	spList := spResponse.GetValue()
+	if len(spList) < 1 {
```
Comment (jakule) [inline lib/integrations/azureoidc/enterprise_app.go:163]: To make sure only one is returned?
```suggestion
	if len(spList) != 1 {
```

File: lib/integrations/azureoidc/private.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+func getPrivateAPIToken(ctx context.Context, tenantID string) (string, error) {
+	tokens, err := getRefreshTokens()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	for _, token := range tokens {
+		slog.DebugContext(ctx, "trying token", "client_id", token.ClientID)
+		token, err := exchangeToken(ctx, tenantID, token)
+		if err != nil {
+			slog.ErrorContext(ctx, "error exchanging token", "err", err)
```
Comment (jakule) [inline lib/integrations/azureoidc/private.go:133]: If incorrect token doesn't break the processing as another one can work this is not an error. I'd lower down the level to warning or even debug. If not token is found, an error is returned from this function anyway.

File: lib/web/integrations_azureoidc.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	}
+
+	integrationName := queryParams.Get("integrationName")
+	if len(integrationName) == 0 {
+		return nil, trace.BadParameter("integrationName must be specified")
+	}
+	// Ensure the integration name is valid.
+	_, err = h.GetProxyClient().GetIntegration(ctx, integrationName)
+	// NotFound error is ignored to prevent disclosure of whether the integration exists in a public/no-auth endpoint.
+	if err != nil && !trace.IsNotFound(err) {
```
Comment (jakule) [inline lib/web/integrations_azureoidc.go:64]: nit: You can skip `err != nil` as `nil` returns false from `trace.IsNotFound()`.

File: tool/teleport/common/integration_configure.go [hunk 7]
```diff
@@ -218,6 +220,32 @@ func onIntegrationConfAccessGraphAWSSync(ctx context.Context, params config.Inte
...
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	fmt.Println()
+	fmt.Println("Success! Use the following information to finish the integration onboarding in Teleport:")
+	fmt.Printf("Tenant ID: %s\nClient ID: %s\n", tenantID, appID)
+	if params.AccessGraph {
+		fmt.Println("Because Access Graph integration was enabled, we also created a file with the necessary data.")
```
Comment (jakule) [inline tool/teleport/common/integration_configure.go:243]: I don't think that the fact that access graph is enabled is important. The important part is that something else needs to be done to finish the setup.
```suggestion
		fmt.Println("To finish the setup you will need `cache.json` file that we created for you.")
```

File: lib/web/integrations_azureoidc.go (hunk 6 above)
Comment (justinas) [reply lib/web/integrations_azureoidc.go:64]: `trace.IsNotFound(nil)` would return false, so `!trace.IsNotFound(nil)` would be true, and we'd return the error. But we don't want to early-return if the error is `nil`.

File: lib/config/configuration.go [hunk 8]
```diff
@@ -263,6 +267,22 @@ type IntegrationConfAccessGraphAWSSync struct {
...
+	// It is used as the OIDC issuer URL, as well as for SAML URIs.
+	ProxyPublicAddr string
+
+	// AuthConnectorName is the name of the SAML connector that will be created on Teleport side.
+	AuthConnectorName string
+
+	// AccessGraph is a flag indicating that access graph integration is requested.
+	// When this is true, the integration script will produce
+	// a cache file necessary for TAG synchronization.
+	AccessGraph bool
```
Comment (jakule) [inline lib/config/configuration.go:283]: nit: I'd rename it to `AccessGraphEnabled` or something like that, but I also don't mind the current version.

File: lib/integrations/azureoidc/accessgraph.go [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+	for _, app := range appResp.GetValue() {
+		appID := app.GetAppId()
+		if appID == nil {
+			slog.WarnContext(ctx, "app ID is nil", "app", app)
+			continue
+		}
+		sp, err := graphClient.ServicePrincipalsWithAppId(appID).Get(ctx, nil)
+		if err != nil {
+			slog.ErrorContext(ctx, "could not retrieve service principal", "app_id", *appID, "error", err)
```
Comment (jakule) [inline lib/integrations/azureoidc/accessgraph.go:102]: This looks like a bug. If the error is not nil we should not try to call `GetId()` a few lines below.

File: lib/integrations/azureoidc/accessgraph.go [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+		federatedSSOV2, err := privateAPIGet(ctx, token, path.Join("ApplicationSso", *spID, "FederatedSsoV2"))
+		if err != nil {
+			slog.WarnContext(ctx, "getting federated SSO v2 info failed", "error", err)
+			continue
+		}
+
+		federatedSSOV2Compressed, err := gzipBytes(federatedSSOV2)
+		if err != nil {
+			slog.WarnContext(ctx, "can not compress the FederatedSsoV2 payload", "error", err)
```
Comment (jakule) [inline lib/integrations/azureoidc/accessgraph.go:127]: The same as above. Is it fine to use `federatedSSOV2Compressed` if an error is returned?

File: lib/integrations/azureoidc/private.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+		Path:   path.Join(tenantID, "oauth2/v2.0/token"),
+		Scheme: "https",
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, uri.String(), strings.NewReader(params.Encode()))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	resp, err := (&http.Client{}).Do(req)
```
Comment (jakule) [inline lib/integrations/azureoidc/private.go:96]: Use `defaults.HTTPClient()` to use client with some "sane" defaults.

File: lib/integrations/azureoidc/private.go [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+		Path:   path.Join("api", endpoint),
+	}
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, uri.String(), nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", accessToken))
+	req.Header.Add("x-ms-client-request-id", uuid.NewString())
+
+	resp, err := (&http.Client{}).Do(req)
```
Comment (jakule) [inline lib/integrations/azureoidc/private.go:157]: Please use `defaults.HTTPClient()` to use client with custom config

**Review by jakule — APPROVED** [review]
Approved with a few blocks to not block.

**Comment by jakule** [comment]
@tigrato Can you please take a look?

**Review by tigrato — APPROVED** [review]

**Pull request merged**